semantics/src/main/kotlin/com/strumenta/kolasu/semantics/symbol/description/SymbolDescription.kt

```diff
@@ -3,46 +3,173 @@ package com.strumenta.kolasu.semantics.symbol.description
 import com.strumenta.kolasu.language.KolasuLanguage
 import com.strumenta.kolasu.model.Named
 import com.strumenta.kolasu.model.Node
+import com.strumenta.kolasu.model.ReferenceByName
+import kotlin.reflect.KClass
+import kotlin.reflect.full.allSuperclasses
 
+/**
+ * The description of a [Node] from the outside,
+ * i.e. from other trees not containing the node.
+ * @property name the name of the symbol
+ * @property identifier the identifier of the described node
+ * @property type a description of the type of the described node
+ * @property fields the visible fields of the described node
+ *
+ * @author Lorenzo Addazi <lorenzo.addazi@strumenta.com>
+ **/
 data class SymbolDescription(
-    override val name: String,
-    val identifier: String,
-    val types: List<String>,
+    override var name: String,
+    var identifier: String,
+    val type: TypeDescription,
     val fields: Map<String, ValueDescription>
 ) : Node(), Named
 
+/**
+ * The type description of a symbol.
+ * @property name the name of the type
+ * @property superTypes possible super types
+ *
+ * @author Lorenzo Addazi <lorenzo.addazi@strumenta.com>
+ **/
+data class TypeDescription(
+    override val name: String,
+    val superTypes: MutableList<TypeDescription> = mutableListOf()
+) : Node(), Named {
+
+    /**
+     * Checks whether the described type is a subtype of the given [type].
+     *
+     * @param type the possible super-type
+     * @return true if [type] is a super-type, false otherwise
+     **/
+    fun isSubTypeOf(type: KClass<*>?): Boolean {
+        return this.name == type?.qualifiedName ||
+            this.superTypes.any { it.isSubTypeOf(type) }
+    }
+
+    /**
+     * Checks whether the described type is a subtype of the given [type].
+     *
+     * @param type the possible super-type
+     * @return true if [type] is a super-type, false otherwise
+     **/
+    fun isSubTypeOf(type: TypeDescription?): Boolean {
+        return this.name == type?.name ||
+            this.superTypes.any { it.isSubTypeOf(type) }
+    }
+
+    /**
+     * Checks whether the described type is a supertype of the given [type].
+     *
+     * @param type the possible subtype
+     * @return true if [type] is a subtype, false otherwise
+     **/
+    fun isSuperTypeOf(type: KClass<*>?): Boolean {
+        return this.name == type?.qualifiedName ||
+            type?.allSuperclasses?.any { this.isSuperTypeOf(it) } ?: false
+    }
+
+    /**
+     * Checks whether the described type is a supertype of the given [type].
+     *
+     * @param type the possible subtype
+     * @return true if [type] is a subtype, false otherwise
+     **/
+    fun isSuperTypeOf(type: TypeDescription?): Boolean {
```

Comment:
Can't we have rules for super typing and sub typing that are language specific? For example generics, variance, contra-variance, etc.?